### ISSUE TITLE
npm update -g causes EMISSINGARG

### ISSUE THREAD

**Issue by michaelkrone** [npm/npm#9659]
Hi.
When trying to update my global modules I get an EMISSINGARG error:
```
$ npm update -g
npm ERR! Linux 3.13.0-63-generic
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "update" "-g"
npm ERR! node v4.1.0
npm ERR! npm  v3.3.3
npm ERR! code EMISSINGARG

npm ERR! typeerror Error: Missing required argument #1
npm ERR! typeerror     at module.exports (/usr/local/lib/node_modules/npm/lib/install/logical-tree.js:12:3)
npm ERR! typeerror     at /usr/local/lib/node_modules/npm/lib/outdated.js:77:16
npm ERR! typeerror     at /usr/local/lib/node_modules/npm/lib/outdated.js:61:20
npm ERR! typeerror     at /usr/local/lib/node_modules/npm/node_modules/read-package-tree/rpt.js:188:9
npm ERR! typeerror     at f (/usr/local/lib/node_modules/npm/node_modules/once/once.js:17:25)
npm ERR! typeerror     at /usr/local/lib/node_modules/npm/node_modules/read-package-tree/rpt.js:154:20
npm ERR! typeerror     at f (/usr/local/lib/node_modules/npm/node_modules/once/once.js:17:25)
npm ERR! typeerror     at then (/usr/local/lib/node_modules/npm/node_modules/read-package-tree/rpt.js:123:22)
npm ERR! typeerror     at /usr/local/lib/node_modules/npm/node_modules/read-package-tree/rpt.js:83:20
npm ERR! typeerror     at fs.js:1579:23
```

Any Ideas on this?

**Comment by genesem** [comment]
npm ERR! Windows_NT 6.1.7601
npm ERR! argv "XXXX\\node.exe" "XXXX\\AppData\\Roaming\\npm\\node_modules\\npm\\bin\\npm-cli.js" "update" "-g"
npm ERR! node v4.1.0
npm ERR! npm  v3.3.3
npm ERR! code ENOSELF

npm ERR! Refusing to install npm as a dependency of itself

**Comment by othiym23** [comment]
This is mos. def a bug, but I think these may be two separate bugs, because @genesem's issue looks unrelated and I'm a little unclear as to what's happening there. For the first issue, https://github.com/npm/npm/blob/master/lib/outdated.js#L61 is getting hit, which is calling https://github.com/npm/npm/blob/master/lib/outdated.js#L76, but https://github.com/npm/npm/blob/master/lib/outdated.js#L77 is expecting there to be a tree object there (and there isn't, because this is an error case), so either there needs to be an error guard before that line, or some kind of defaulting behavior in the absence of `physicalTree`. @iarna, can you take a look at this?

**Comment by iamstarkov** [comment]
facing the same problem right now

**Comment by genesem** [comment]
@iamstarkov I solved this only with manual install  of npm.

Download it from here: https://github.com/npm/npm/releases
direct link: https://github.com/npm/npm/archive/v3.3.6.zip

Unzip and copy everything to global `node_modules/npm`
on my system it's: `C:\Users\gs\AppData\Roaming\npm\node_modules\npm`

**Comment by iarna** [comment]
@michaelkrone: Can you give this a try with `npm@3.3.6`? I'm pretty sure we fixed it a bit ago.

@genesem: Your issue was _definitely_ fixed in 3.3.5. You could also work around it with `npm install -f -g npm`.

@iamstarkov There are two different issues discussed here, which one is biting you?

**Comment by genesem** [comment]
Thank you guys, `npm i npm -g -f` solved the issue.

It was requred to set congig for global packages after reinstall with command:
`npm config set prefix "${APPDATA}/npm" -g`

**Comment by iamstarkov** [comment]
@iarna mine issue is about unable to update npm facing EMISSINGARG error

**Comment by genesem** [comment]
@iamstarkov  type `npm i npm -g -f` in console.

**Comment by iamstarkov** [comment]
@iarna I tried `npm up -gf npm`  and it worked, after it typed `npm config set prefix "${APPDATA}/npm" -g` and it broke npm again:

```
➜  ~   npm config set prefix "${APPDATA}/npm" -g
➜  ~  npm up -g npm
Error: EACCES: permission denied, mkdir '/npm'
    at Error (native)
```

**Comment by genesem** [comment]
@iamstarkov this `pm config set prefix "${APPDATA}/npm" -g` *only* for windows.
remove all and install it again the best choice.

**Comment by cigaly** [comment]
I've reinstalled np, now is v3.3.6 ... and the problem is still present :-(

**Comment by iamstarkov** [comment]
@genesem is there any way to redo `npm config set prefix "${APPDATA}/npm" -g`?

**Comment by iarna** [comment]
@iamstarkov You want to run `npm config delete prefix -g` to clear that

**Comment by iarna** [comment]
Since everyone seems sorted out, I'm gonna close this issue. If you all run into something new, please feel free to open a new issue. =)

**Comment by michaelkrone** [comment]
@iarna sorry for the late response. The error persists with npm@3.3.6:

```
0 info it worked if it ends with ok
1 verbose cli [ '/usr/local/bin/node', '/usr/local/bin/npm', 'update', '-g' ]
2 info using npm@3.3.6
3 info using node@v4.1.2
4 verbose stack Error: Missing required argument #1
4 verbose stack     at module.exports (/usr/local/lib/node_modules/npm/lib/install/mutate-into-logical-tree.js:11:3)
4 verbose stack     at /usr/local/lib/node_modules/npm/lib/outdated.js:77:5
4 verbose stack     at /usr/local/lib/node_modules/npm/lib/outdated.js:61:20
4 verbose stack     at /usr/local/lib/node_modules/npm/node_modules/read-package-tree/rpt.js:186:9
4 verbose stack     at f (/usr/local/lib/node_modules/npm/node_modules/once/once.js:17:25)
4 verbose stack     at /usr/local/lib/node_modules/npm/node_modules/read-package-tree/rpt.js:152:20
4 verbose stack     at f (/usr/local/lib/node_modules/npm/node_modules/once/once.js:17:25)
4 verbose stack     at then (/usr/local/lib/node_modules/npm/node_modules/read-package-tree/rpt.js:121:22)
4 verbose stack     at /usr/local/lib/node_modules/npm/node_modules/read-package-tree/rpt.js:81:20
4 verbose stack     at fs.js:1579:23
5 verbose cwd /home/michael/dev/instruments/gastro-key
6 error Linux 3.13.0-63-generic
7 error argv "/usr/local/bin/node" "/usr/local/bin/npm" "update" "-g"
8 error node v4.1.2
9 error npm  v3.3.6
10 error code EMISSINGARG
11 error typeerror Error: Missing required argument #1
11 error typeerror     at module.exports (/usr/local/lib/node_modules/npm/lib/install/mutate-into-logical-tree.js:11:3)
11 error typeerror     at /usr/local/lib/node_modules/npm/lib/outdated.js:77:5
11 error typeerror     at /usr/local/lib/node_modules/npm/lib/outdated.js:61:20
11 error typeerror     at /usr/local/lib/node_modules/npm/node_modules/read-package-tree/rpt.js:186:9
11 error typeerror     at f (/usr/local/lib/node_modules/npm/node_modules/once/once.js:17:25)
11 error typeerror     at /usr/local/lib/node_modules/npm/node_modules/read-package-tree/rpt.js:152:20
11 error typeerror     at f (/usr/local/lib/node_modules/npm/node_modules/once/once.js:17:25)
11 error typeerror     at then (/usr/local/lib/node_modules/npm/node_modules/read-package-tree/rpt.js:121:22)
11 error typeerror     at /usr/local/lib/node_modules/npm/node_modules/read-package-tree/rpt.js:81:20
11 error typeerror     at fs.js:1579:23
11 error typeerror This is an error with npm itself. Please report this error at:
11 error typeerror     <http://github.com/npm/npm/issues>
12 verbose exit [ 1, true ]
```

**Comment by iamstarkov** [comment]
@iarna npm seems to be completely broken now and `config delete` is not working:
```
➜  postcss-color-mix git:(master) npm config delete prefix -g
Error: EACCES: permission denied, mkdir '/npm'
    at Error (native)
```

**Comment by iamstarkov** [comment]
fixed it 

```
➜  postcss-color-mix git:(master) sudo chown -R $USER /npm
Password:
chown: /npm: No such file or directory
➜  postcss-color-mix git:(master) mkdir /npm
mkdir: /npm: Permission denied
➜  postcss-color-mix git:(master) sudo mkdir /npm
➜  postcss-color-mix git:(master) sudo chown -R $USER /npm
➜  postcss-color-mix git:(master) npm config delete prefix -g
➜  postcss-color-mix git:(master) npm update -g npm
…
npm@3.3.6 /usr/local/lib/node_modules/npm
➜  postcss-color-mix git:(master) npm -v
3.3.6
```

**Comment by ErikAbele** [comment]
Still getting this error when running:

```
[erik@odin ~]% npm update npm -g

npm ERR! Darwin 12.6.0
npm ERR! argv "/usr/local/Cellar/node/4.2.1/bin/node" "/usr/local/bin/npm" "update" "npm" "-g"
npm ERR! node v4.2.1
npm ERR! npm  v3.3.9
npm ERR! code EMISSINGARG

npm ERR! typeerror Error: Missing required argument #1
npm ERR! typeerror     at module.exports (/usr/local/lib/node_modules/npm/lib/install/mutate-into-logical-tree.js:12:3)
npm ERR! typeerror     at /usr/local/lib/node_modules/npm/lib/outdated.js:78:5
npm ERR! typeerror     at /usr/local/lib/node_modules/npm/lib/outdated.js:62:20
npm ERR! typeerror     at /usr/local/lib/node_modules/npm/node_modules/read-package-tree/rpt.js:186:9
npm ERR! typeerror     at f (/usr/local/lib/node_modules/npm/node_modules/once/once.js:17:25)
npm ERR! typeerror     at /usr/local/lib/node_modules/npm/node_modules/read-package-tree/rpt.js:152:20
npm ERR! typeerror     at f (/usr/local/lib/node_modules/npm/node_modules/once/once.js:17:25)
npm ERR! typeerror     at then (/usr/local/lib/node_modules/npm/node_modules/read-package-tree/rpt.js:121:22)
npm ERR! typeerror     at /usr/local/lib/node_modules/npm/node_modules/read-package-tree/rpt.js:81:20
npm ERR! typeerror     at fs.js:1583:23
npm ERR! typeerror This is an error with npm itself. Please report this error at:
npm ERR! typeerror     <http://github.com/npm/npm/issues>

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/erik/npm-debug.log
Exit 1
```

**Comment by ErikAbele** [comment]
Sorry, GitHub refuses to accept the npm-debug.log (even when renamed to .txt).

**Comment by yosiat** [comment]
I am getting this as well, on every global command (update -g, outdated -g):

```
λ ~/ npm outdated -g
npm ERR! Darwin 15.0.0
npm ERR! argv "/usr/local/Cellar/node/5.1.0/bin/node" "/usr/local/bin/npm" "outdated" "-g"
npm ERR! node v5.1.0
npm ERR! npm  v3.5.0
npm ERR! code EMISSINGARG

npm ERR! typeerror Error: Missing required argument #1
npm ERR! typeerror     at module.exports (/usr/local/lib/node_modules/npm/lib/install/mutate-into-logical-tree.js:12:3)
npm ERR! typeerror     at /usr/local/lib/node_modules/npm/lib/outdated.js:78:5
npm ERR! typeerror     at /usr/local/lib/node_modules/npm/lib/outdated.js:62:20
npm ERR! typeerror     at /usr/local/lib/node_modules/npm/node_modules/read-package-tree/rpt.js:186:9
npm ERR! typeerror     at f (/usr/local/lib/node_modules/npm/node_modules/once/once.js:17:25)
npm ERR! typeerror     at /usr/local/lib/node_modules/npm/node_modules/read-package-tree/rpt.js:152:20
npm ERR! typeerror     at f (/usr/local/lib/node_modules/npm/node_modules/once/once.js:17:25)
npm ERR! typeerror     at then (/usr/local/lib/node_modules/npm/node_modules/read-package-tree/rpt.js:121:22)
npm ERR! typeerror     at /usr/local/lib/node_modules/npm/node_modules/read-package-tree/rpt.js:81:20
npm ERR! typeerror     at fs.js:1657:23
npm ERR! typeerror This is an error with npm itself. Please report this error at:
npm ERR! typeerror     <http://github.com/npm/npm/issues>
```

**Comment by eisisig** [comment]
I was having this same problem (Missing argument). Nothing worked. Updated npm and node. I ended up uninstalling node and removing /usr/local/lib/node_modules which had the global packages... After installing node and npm again, "npm update -g" works

I'm on mac and using brew for node

**Comment by iamstarkov** [comment]
did you have sym links in you node_modules?

**Comment by eisisig** [comment]
@iamstarkov Yes!

**Comment by iamstarkov** [comment]
yeah, basically removing global node modules sym links helped few times

**Comment by eisisig** [comment]
@iamstarkov Thats good to know. I'll give it more time to debug if this starts to happen again

**Comment by amandeepmittal** [comment]
I am getting the error: "argv "/usr/local/bin/node" "/usr/local/bin/npm" "install" "jade"
"Refusing to install jade as a dependency of itself"
with npm version 3.3.12
Any thoughts?

**Comment by iamstarkov** [comment]
@amandeepmittal 

> "Refusing to install jade as a dependency of itself"

this maybe?

**Comment by eisisig** [comment]
@amandeepmittal You get this when you try to install a package with the same name as defined in the current folders package.json

So maybe you are inside the jade package folder?

**Comment by amandeepmittal** [comment]
@iamstarkov @eisisig No, I wan't inside the Jade Package folder, I was inside my applications root folder. Well I got the solution to my problem by installing Jade npm first and doing an "npm init" second. Not sure why it worked, or if that's how it should work but the main problem was the package.son file the "npm init" command had create before.

**Comment by theofidry** [comment]
Hello,

I have the same issue both locally (OSX 10.9) and on Travis (cf. [build](https://travis-ci.org/theofidry/travis-bug-demo/jobs/128909465)).

- node: `4.2.6`
- npm: `3.9.0`

Short message:

<img width="877" alt="screen shot 2016-05-09 at 18 23 09" src="https://cloud.githubusercontent.com/assets/5175937/15121559/1e0b4210-1613-11e6-990e-a5c3d5f5186a.png">

Full `npm-debug.log` content [here](https://gist.github.com/theofidry/3212be0e1c08bddfae928360820cfce9).

**Comment by theofidry** [comment]
In my case downgrading npm to `3.8.9` (instead of previously `3.9.0`) solved my issue... Hope it helps.

**Comment by joewood** [comment]
I discovered a reason for this error:

* I had an `npm link` package used on my system. The linked package itself had another package also linked on the system. 
* I had previously performed a `npm unlink <subpackage>` inside the linked package to remove the subpackage.
* After I did this I started seeing this error.

**npm** didn't seem to like the state of the package after the `npm unlink`. The reason for the problem was that the `unlink` had left the `node_modules/subpackage` directory after the unlink, and the directory was empty. As others had previously mentioned, **npm** doesn't like a linked package without a package.json file. This seems to also apply to any subpackages in those linked packages.

So, if you see this error check all your `npm link`ed packages for any empty package folders under `node_modules`.
